lib/task.ml

```diff
@@ -1,33 +1,102 @@
+open EffectHandlers
+open EffectHandlers.Deep
+
 type 'a task = unit -> 'a
 
-type 'a promise = ('a, exn) result option Atomic.t
+type 'a promise_state =
+  Returned of 'a
+| Raised of exn
+| Pending of (unit, unit) continuation list
 
-exception TasksActive
+type 'a promise = 'a promise_state Atomic.t
 
-type task_msg =
-  Task : 'a task * 'a promise -> task_msg
-| Quit : task_msg
+type message =
+  Work of (unit -> unit)
+| Quit
 
 type pool_data = {
   domains : unit Domain.t array;
-  task_chan : task_msg Multi_channel.t;
+  task_chan : message Multi_channel.t;
   name: string option
 }
 
 type pool = pool_data option Atomic.t
 
-let do_task f p =
+let get_pool_data p =
+  match Atomic.get p with
+  | None -> raise (Invalid_argument "pool already torn down")
+  | Some p -> p
+
+let do_task pd f p =
   try
     let res = f () in
-    Atomic.set p (Some (Ok res))
+    match Atomic.exchange p (Returned res) with
+    | Pending l ->
+        List.iter (fun k -> Multi_channel.send pd.task_chan (Work (continue k))) l
+    | _ -> failwith "Task.do_task: impossible (1)"
   with e ->
-    Atomic.set p (Some (Error e));
-    match e with
-    | TasksActive -> raise e
-    | _ -> ()
+    begin match Atomic.exchange p (Raised e) with
+    | Pending l -> List.iter (fun k ->
+        Multi_channel.send pd.task_chan (Work (fun _ -> discontinue k e))) l
+    | _ -> failwith "Task.do_task: impossible (2)"
+    end
 
-let named_pools = Hashtbl.create 8
+let async pool f =
+  let pd = get_pool_data pool in
+  let p = Atomic.make (Pending []) in
+  Multi_channel.send pd.task_chan (Work (fun _ -> do_task pd f p));
+  p
+
+type _ eff += Wait : 'a promise -> unit eff
 
+let rec await promise =
+  match Atomic.get promise with
+  | Returned v -> v
+  | Raised e -> raise e
```

Comment:
Should we save the raw_backtrace and use raise_with_backtrace here? The backtrace might look all wrong if it is just raised like this here (is it per process or per domain now?)

Reply:
This needs a variant of `discontinue` that can raise with `raw_backtrace` in the compiler. I'll add support for that in the compiler.